helix-core/src/main/java/org/apache/helix/controller/stages/PerReplicaThrottleStage.java

```diff
@@ -0,0 +1,270 @@
+      for (Partition partition : selectedResourceMessages.keySet()) {
+        output.addMessages(resourceName, partition, selectedResourceMessages.get(partition));
+      }
+      return;
+    }
+
+    Map<Partition, Map<String, Integer>> expectedStateCountByPartition = new HashMap<>();
+    Map<Partition, Map<String, Integer>> currentStateCountsByPartition = new HashMap<>();
+    // TODO: later PRs
+    // Step 1: charge existing pending messages and update retraced state map.
+    // Step 2: classify all the messages into recovery message list and load message list
+    // Step 3: sorts recovery message list and applies throttling
+    // Step 4: sorts load message list and applies throttling
+    // Step 5: construct output
+    Map<Partition, List<Message>> outMessagesByPartition = new HashMap<>();
+    for (Partition partition : resource.getPartitions()) {
+      List<Message> partitionMessages = selectedResourceMessages.get(partition);
+      if (partitionMessages == null) {
+        continue;
+      }
+      List<Message> finalPartitionMessages = new ArrayList<>();
+      for (Message message: partitionMessages) {
+        // TODO: next PR messages exclusion
+        finalPartitionMessages.add(message);
+      }
+      outMessagesByPartition.put(partition, finalPartitionMessages);
+      output.addMessages(resourceName, partition, finalPartitionMessages);
+    }
+    // Step 6: constructs all retraced partition state map for the resource
+    // TODO: next PR
+    // Step 7: emit metrics
+    // TODO: next PR
+  }
+
+
+  // ------------------ utilities ---------------------------
+  /**
+   * POJO that maps resource name to its priority represented by an integer.
+   */
+  private static class ResourcePriority implements Comparable<ResourcePriority> {
+    private String _resourceName;
+    private int _priority;
+
+    ResourcePriority(String resourceName, ResourceControllerDataProvider dataCache) {
+      // Resource level prioritization based on the numerical (sortable) priority field.
+      // If the resource priority field is null/not set, the resource will be treated as lowest
+      // priority.
+      _priority = Integer.MIN_VALUE;
+      _resourceName = resourceName;
+      String priorityField = dataCache.getClusterConfig().getResourcePriorityField();
+      if (priorityField != null) {
+        // Will take the priority from ResourceConfig first
+        // If ResourceConfig does not exist or does not have this field.
+        // Try to load it from the resource's IdealState. Otherwise, keep it at the lowest priority
+        if (dataCache.getResourceConfig(resourceName) != null
+            && dataCache.getResourceConfig(resourceName).getSimpleConfig(priorityField) != null) {
+          this.setPriority(
+              dataCache.getResourceConfig(resourceName).getSimpleConfig(priorityField));
+        } else if (dataCache.getIdealState(resourceName) != null
```

Comment:
nit, these codes could be improved a little bit, IMHO. Note the "dataCache.getResourceConfig(resourceName)" call is made 3 times.

I suggest changing to the following,
ResourceConfig config = dataCache.getResourceConfig(resourceName)
if (config != null) {
  String priority = dataCache.getResourceConfig(resourceName).getSimpleConfig(priorityField);
  this.setPriority(priority == null ? DEFAULT_PRIORITY : priority);
}

The other condition branch can be simplified in the same way.

Reply:
1. We can always add the string.
2. The conversion is happening anyway. The value you get from the Map is also a String. This is a different topic.
3. Why not make this small change to simplify code without any side-effect?

Reply:
Changed to 

```
      _priority = Integer.valueOf(DEFAULT_PRIORITY);
      _resourceName = resourceName;
      String priorityField = dataCache.getClusterConfig().getResourcePriorityField();

      if (priorityField != null) {
        // Will take the priority from ResourceConfig first
        // If ResourceConfig does not exist or does not have this field.
        // Try to load it from the resource's IdealState. Otherwise, keep it at the lowest priority
        ResourceConfig config = dataCache.getResourceConfig(resourceName);
        IdealState idealState = dataCache.getIdealState(resourceName);
        if (config != null) {
          String priority = dataCache.getResourceConfig(resourceName).getSimpleConfig(priorityField);
          this.setPriority(priority == null ? DEFAULT_PRIORITY : priority);
        } else if (idealState != null) {
          String priority = idealState.getRecord().getSimpleField(priorityField);
          this.setPriority(priority == null ? DEFAULT_PRIORITY : priority);
        }
      }
```

Reply:
On a second thought, @jiajunwang , I think there is a bug in your new proposed code. They are not equivalent to the original code. 

Assuming the above is what you suggested. The case that if resourceConfig having priorityField, but retrieved value priority is null, we have priority as min_int (DEFAULT_PRIORITY). However, in the original code, in the original code, we should check idealstate priorityField value. 

This is kind of tricky, as simplifying the original code has change to introduce new bug.  since the above code is pushed to  this PR now. Shall we use the original code or shall we use following?

```
 private final String DEFAULT_PRIORITY = Integer.valueOf(Integer.MIN_VALUE).toString();

      _priority = Integer.valueOf(DEFAULT_PRIORITY);
      _resourceName = resourceName;
      String priorityField = dataCache.getClusterConfig().getResourcePriorityField();

      if (priorityField != null) {
        // Will take the priority from ResourceConfig first
        // If ResourceConfig does not exist or does not have this field.
        // Try to load it from the resource's IdealState. Otherwise, keep it at the lowest priority
        ResourceConfig config = dataCache.getResourceConfig(resourceName);
        IdealState idealState = dataCache.getIdealState(resourceName);
        if (config != null) {
          String priority = dataCache.getResourceConfig(resourceName).getSimpleConfig(priorityField);
          if (priority != null) {
              this.setPriority priority);
          }
        } else if (idealState != null) {
          String priority = idealState.getRecord().getSimpleField(priorityField);
          this.setPriority(priority == null ? DEFAULT_PRIORITY : priority);
        }
      }

```

Reply:
Sync-ed with JJ offline, the main concern is `getResourceConfig` is used three time, better to use once. Same for `getIdealState`. 
Reduce to using once would work, still keeping the original code logic.